### PR TITLE
Optimize integer arithmetic

### PR DESCRIPTION
**Pull request by bjorng**
This pull request optimizes some of the arithmetic operations for integers.

The most noticeable improvement is in multiplication of large integers. Consider this benchmark:

```
bench() ->
    Size = 1_262_000,
    String = binary:copy(<<"9">>, Size),
    {Time, _Val} = timer:tc(erlang, binary_to_integer, [String]),
    io:format("Size: ~p, seconds: ~p\n", [Size, Time / 1_000_000]).
```

Its running time is dominated by multiplication of large integers. This PR reduces the running time from about 0.45 seconds down to about 0.32 seconds on my computer (an M1 MacBook Pro).

### CODE REVIEW

File: erts/emulator/beam/jit/arm/beam_asm.hpp [hunk 1]
```diff
@@ -1128,6 +1128,15 @@ class BeamModuleAssembler : public BeamAssembler,
                              const a64::Gp rhs_reg,
                              const Label next);
 
+    void emit_div_rem_literal(Sint divisor,
+                              const ArgSource &Dividend,
+                              arm::Gp dividend,
+                              arm::Gp quotient,
+                              arm::Gp remainder,
+                              const Label generic,
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/beam_asm.hpp:1136]:
```suggestion
                              const Label &generic,
```

File: erts/emulator/beam/jit/arm/instr_arith.cpp [hunk 2]
```diff
@@ -673,6 +853,95 @@ void BeamGlobalAssembler::emit_int_div_rem_body_shared() {
     }
 }
 
+void BeamModuleAssembler::emit_div_rem_literal(Sint divisor,
+                                               const ArgSource &Dividend,
+                                               arm::Gp dividend,
+                                               arm::Gp quotient,
+                                               arm::Gp remainder,
+                                               const Label generic,
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_arith.cpp:861]:
```suggestion
                                               const Label &generic,
```

File: erts/emulator/beam/jit/arm/instr_arith.cpp [hunk 3]
```diff
@@ -685,59 +954,51 @@ void BeamModuleAssembler::emit_div_rem(const ArgLabel &Fail,
 
     if (RHS.isSmall()) {
         divisor = RHS.as<ArgSmall>().getSigned();
+        if (divisor == (Sint)-1) {
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_arith.cpp:957]: Shouldn't the integer promotion rules make this cast redundant?

File: erts/emulator/beam/jit/arm/instr_arith.cpp [hunk 4]
```diff
@@ -1221,34 +1483,57 @@ void BeamModuleAssembler::emit_i_bsr(const ArgLabel &Fail,
...
+                a.cmp(small_tag, imm(_TAG_IMMED1_SMALL));
                 a.b_ne(generic);
             }
 
             /* We don't need to clear the mask after shifting because
              * _TAG_IMMED1_SMALL will set all the bits anyway. */
             ERTS_CT_ASSERT(_TAG_IMMED1_MASK == _TAG_IMMED1_SMALL);
-            a.asr(TMP1, lhs.reg, imm(shift));
-            a.orr(dst.reg, TMP1, imm(_TAG_IMMED1_SMALL));
+            shift = std::min(shift, (Sint)63);
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_arith.cpp:1507]:
```suggestion
            shift = std::min<Sint>(shift, 63);
```

File: erts/emulator/beam/jit/x86/instr_arith.cpp [hunk 5]
```diff
@@ -841,115 +900,263 @@ void BeamGlobalAssembler::emit_times_guard_shared() {
...
+        factor = Src2.as<ArgSmall>().getSigned();
+        if (Support::isPowerOf2(factor)) {
+            left_shift = Support::ctz<Eterm>(factor);
+        }
+    }
+
+    if (always_small(Src1) && Src2.isSmall() && Src4.isSmall() &&
+        is_product_small && is_sum_small) {
+        x86::Mem p;
+        Sint increment = Src4.as<ArgSmall>().get();
```
Comment (jhogberg) [inline erts/emulator/beam/jit/x86/instr_arith.cpp:1017]: If I'm not mistaken, we can skip untagging by letting `increment` cancel out the multiplied tag bits.

File: erts/emulator/beam/jit/x86/instr_arith.cpp [hunk 6]
```diff
@@ -1325,21 +1533,41 @@ void BeamModuleAssembler::emit_i_bsr(const ArgSource &LHS,
             /* We don't need to clear the mask after shifting because
              * _TAG_IMMED1_SMALL will set all the bits anyway. */
             ERTS_CT_ASSERT(_TAG_IMMED1_MASK == _TAG_IMMED1_SMALL);
+            shift = std::min(shift, (Sint)63);
```
Comment (jhogberg) [inline erts/emulator/beam/jit/x86/instr_arith.cpp:1536]:
```suggestion
            shift = std::min<Sint>(shift, 63);
```

**Review by jhogberg — APPROVED** [review]
LGTM

File: erts/emulator/beam/jit/x86/instr_arith.cpp (hunk 5 above)
Comment (bjorng) [reply erts/emulator/beam/jit/x86/instr_arith.cpp:1017]: Yes, that should work. I'll try to do it on Monday.

Comment (bjorng) [reply erts/emulator/beam/jit/x86/instr_arith.cpp:1017]: I've pushed a commit that skips untagging.

File: erts/emulator/beam/jit/arm/instr_arith.cpp [hunk 7]
```diff
@@ -1519,11 +1519,18 @@ void BeamModuleAssembler::emit_i_bsr(const ArgLabel &Fail,
...
-        a.b_lt(generic);
+        if (std::get<0>(getClampedRange(RHS)) >= 0) {
+            a.and_(TMP1, TMP1, imm(_TAG_IMMED1_MASK));
+            a.cmp(TMP1, imm(_TAG_IMMED1_SMALL));
+            a.b_ne(generic);
+        } else {
+            a.mvn(TMP1, TMP1);
+            a.tst(TMP1, imm(_TAG_IMMED1_MASK));
+            a.ccmp(rhs.reg, imm(0), imm(NZCV::kSigned), arm::CondCode::kEQ);
+            a.b_lt(generic);
```
Comment (jhogberg) [inline erts/emulator/beam/jit/arm/instr_arith.cpp:1532]: Slightly shorter, assuming we move the initial `a.and_(TMP1, lhs.reg, rhs.reg);` into the other branch:
```suggestion
            a.ands(TMP1, rhs.reg, imm((1ul << 63) | _TAG_IMMED1_SMALL));
            a.and_(TMP1, lhs.reg, TMP1);
            a.ccmp(TMP1, imm(_TAG_IMMED1_SMALL), imm(NZCV::kNone), arm::CondCode::kPL);
            a.b_ne(generic);
```

Comment (bjorng) [reply erts/emulator/beam/jit/arm/instr_arith.cpp:1532]: Nice! I removed the other branch since it is no longer any shorter.

**Pull request merged**